python/ray/_raylet.pyx

```diff
@@ -0,0 +1,647 @@
+                 ObjectID actor_creation_dummy_object_id,
+                 int32_t max_actor_reconstructions, ActorID actor_id,
+                 ActorHandleID actor_handle_id, int actor_counter,
+                 new_actor_handles, execution_arguments, resource_map,
+                 placement_resource_map):
+        cdef:
+            unordered_map[c_string, double] required_resources
+            unordered_map[c_string, double] required_placement_resources
+            c_vector[shared_ptr[RayletTaskArgument]] task_args
+            c_vector[CActorHandleID] task_new_actor_handles
+            c_vector[c_string] c_function_descriptor
+            c_string pickled_str
+            c_vector[CObjectID] references
+
+        for item in function_descriptor:
+            if not isinstance(item, bytes):
+                raise TypeError("'function_descriptor' takes a list of byte strings.")
+            c_function_descriptor.push_back(item)
+
+        # Parse the resource map.
+        if resource_map is not None:
+            required_resources = resource_map_from_python_dict(resource_map)
+        if required_resources.count(b"CPU") == 0:
+            required_resources[b"CPU"] = 1.0
+        if placement_resource_map is not None:
+            required_placement_resources = resource_map_from_python_dict(placement_resource_map)
+
+        # Parse the arguments from the list.
+        for arg in arguments:
+            if isinstance(arg, ObjectID):
+                references = c_vector[CObjectID]()
+                references.push_back((<ObjectID>arg).data)
+                task_args.push_back(static_pointer_cast[RayletTaskArgument, RayletTaskArgumentByReference](make_shared[RayletTaskArgumentByReference](references)))
+            else:
+                pickled_str = pickle.dumps(arg, protocol=pickle.HIGHEST_PROTOCOL)
+                task_args.push_back(static_pointer_cast[RayletTaskArgument, RayletTaskArgumentByValue](make_shared[RayletTaskArgumentByValue](<uint8_t *>pickled_str.c_str(), pickled_str.size())))
+
+        for new_actor_handle in new_actor_handles:
+            task_new_actor_handles.push_back((<ActorHandleID?>new_actor_handle).data)
+
+        self.task_spec.reset(new RayletTaskSpecification(
+            CUniqueID(driver_id.data), parent_task_id.data, parent_counter, actor_creation_id.data,
+            actor_creation_dummy_object_id.data, max_actor_reconstructions, CUniqueID(actor_id.data),
+            CUniqueID(actor_handle_id.data), actor_counter, task_new_actor_handles, task_args, num_returns,
+            required_resources, required_placement_resources, LANGUAGE_PYTHON,
+            c_function_descriptor))
+
+        # Set the task's execution dependencies.
+        self.execution_dependencies.reset(new c_vector[CObjectID]())
+        if execution_arguments is not None:
+            for execution_arg in execution_arguments:
+                self.execution_dependencies.get().push_back((<ObjectID?>execution_arg).data)
```

Comment:
What does `<ObjectID?>` mean?

Reply:
Cython will check and convert the type of the object. If it is not a ObjectID, Cython will raise a TypeError; otherwise Cython will treat it as a Cython `ObjectID` object (it is different from a Python object because we can access its native components and it is statically compiled).